### PR TITLE
webgpu: Fix the potential error using atomicCompareExchangeWeak

### PR DESCRIPTION
**Pull request by qjia7**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5707)

### CODE REVIEW

**Comment by qjia7** [comment]
@hujiajie @lina128 @pyu10055 PTAL, thanks!

**Comment by xhcao** [comment]
LGTM

**Comment by qjia7** [comment]
@lina128 @pyu10055 I need your LGTM to merge this PR. Thanks.

**Review by lina128 — APPROVED** [review]
LGTM

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5707#-:-MnCIf5C3x2rbau8tU6R:bb6ywmx)* status: :shipit: complete! 3 of 1 approvals obtained (waiting on @qjia7)

**Comment by qjia7** [comment]
@lina128 @mattsoulanille It seems that the latest bots have some problems:
```
Step #4 - "bazel-tests": ==================== Test output for //tfjs-converter/python/tensorflowjs/converters:tf_saved_model_conversion_v2_test:
Step #4 - "bazel-tests": 2021-11-01 01:27:27.605874: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
Step #4 - "bazel-tests": 2021-11-01 01:27:27.941946: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
Step #4 - "bazel-tests": Traceback (most recent call last):
Step #4 - "bazel-tests":   File "/builder/home/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/execroot/tfjs/bazel-out/k8-fastbuild/bin/tfjs-converter/python/tensorflowjs/converters/tf_saved_model_conversion_v2_test.runfiles/tfjs/tfjs-converter/python/tensorflowjs/converters/tf_saved_model_conversion_v2_test.py", line 34, in <module>
Step #4 - "bazel-tests":     import tensorflow_hub as hub
Step #4 - "bazel-tests":   File "/builder/home/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/execroot/tfjs/bazel-out/k8-fastbuild/bin/tfjs-converter/python/tensorflowjs/converters/tf_saved_model_conversion_v2_test.runfiles/tensorflowjs_dev_deps_linux_amd64/pypi__tensorflow_hub/tensorflow_hub/__init__.py", line 88, in <module>
Step #4 - "bazel-tests":     from tensorflow_hub.estimator import LatestModuleExporter
Step #4 - "bazel-tests":   File "/builder/home/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/execroot/tfjs/bazel-out/k8-fastbuild/bin/tfjs-converter/python/tensorflowjs/converters/tf_saved_model_conversion_v2_test.runfiles/tensorflowjs_dev_deps_linux_amd64/pypi__tensorflow_hub/tensorflow_hub/estimator.py", line 62, in <module>
Step #4 - "bazel-tests":     class LatestModuleExporter(tf.compat.v1.estimator.Exporter):
Step #4 - "bazel-tests":   File "/builder/home/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/execroot/tfjs/bazel-out/k8-fastbuild/bin/tfjs-converter/python/tensorflowjs/converters/tf_saved_model_conversion_v2_test.runfiles/tensorflowjs_dev_deps_linux_amd64/pypi__tensorflow/tensorflow/python/util/lazy_loader.py", line 62, in __getattr__
Step #4 - "bazel-tests":     module = self._load()
Step #4 - "bazel-tests":   File "/builder/home/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/execroot/tfjs/bazel-out/k8-fastbuild/bin/tfjs-converter/python/tensorflowjs/converters/tf_saved_model_conversion_v2_test.runfiles/tensorflowjs_dev_deps_linux_amd64/pypi__tensorflow/tensorflow/python/util/lazy_loader.py", line 45, in _load
Step #4 - "bazel-tests":     module = importlib.import_module(self.__name__)
Step #4 - "bazel-tests":   File "/builder/home/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/python3_interpreter_linux_amd64/bazel_install_py3/lib/python3.8/importlib/__init__.py", line 127, in import_module
Step #4 - "bazel-tests":     return _bootstrap._gcd_import(name[level:], package, level)
Step #4 - "bazel-tests":   File "/builder/home/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/execroot/tfjs/bazel-out/k8-fastbuild/bin/tfjs-converter/python/tensorflowjs/converters/tf_saved_model_conversion_v2_test.runfiles/tensorflowjs_dev_deps_linux_amd64/pypi__tensorflow_estimator/tensorflow_estimator/__init__.py", line 10, in <module>
Step #4 - "bazel-tests":     from tensorflow_estimator._api.v1 import estimator
Step #4 - "bazel-tests":   File "/builder/home/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/execroot/tfjs/bazel-out/k8-fastbuild/bin/tfjs-converter/python/tensorflowjs/converters/tf_saved_model_conversion_v2_test.runfiles/tensorflowjs_dev_deps_linux_amd64/pypi__tensorflow_estimator/tensorflow_estimator/_api/v1/estimator/__init__.py", line 10, in <module>
Step #4 - "bazel-tests":     from tensorflow_estimator._api.v1.estimator import experimental
Step #4 - "bazel-tests":   File "/builder/home/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/execroot/tfjs/bazel-out/k8-fastbuild/bin/tfjs-converter/python/tensorflowjs/converters/tf_saved_model_conversion_v2_test.runfiles/tensorflowjs_dev_deps_linux_amd64/pypi__tensorflow_estimator/tensorflow_estimator/_api/v1/estimator/experimental/__init__.py", line 10, in <module>
Step #4 - "bazel-tests":     from tensorflow_estimator.python.estimator.canned.dnn import dnn_logit_fn_builder
Step #4 - "bazel-tests":   File "/builder/home/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/execroot/tfjs/bazel-out/k8-fastbuild/bin/tfjs-converter/python/tensorflowjs/converters/tf_saved_model_conversion_v2_test.runfiles/tensorflowjs_dev_deps_linux_amd64/pypi__tensorflow_estimator/tensorflow_estimator/python/estimator/canned/dnn.py", line 27, in <module>
Step #4 - "bazel-tests":     from tensorflow_estimator.python.estimator import estimator
Step #4 - "bazel-tests":   File "/builder/home/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/execroot/tfjs/bazel-out/k8-fastbuild/bin/tfjs-converter/python/tensorflowjs/converters/tf_saved_model_conversion_v2_test.runfiles/tensorflowjs_dev_deps_linux_amd64/pypi__tensorflow_estimator/tensorflow_estimator/python/estimator/estimator.py", line 70, in <module>
Step #4 - "bazel-tests":     @doc_controls.inheritable_header("""\
Step #4 - "bazel-tests": AttributeError: module 'tensorflow.tools.docs.doc_controls' has no attribute 'inheritable_header'

**Comment by lina128** [comment]
Hi @qjia7 , it’s just fixed. Please try again.

File: tfjs-backend-webgpu/src/kernels/scatter_optimized_webgpu.ts [hunk 1]
```diff
@@ -98,13 +98,16 @@ export class ScatterOptimizedProgram implements WebGPUProgram {
...
         `ignore(atomicAdd(&(result.numbers[flatIndex]), i32(updateValue)));` :
         `
-     var oldI32 = atomicLoad(&(result.numbers[flatIndex]));
-     var assumed = oldI32 - 1;
-     for (; assumed != oldI32;) {
-       assumed = oldI32;
+     var assumed = atomicLoad(&(result.numbers[flatIndex]));
+     var success = 0;
+     // TODO: Will it have the risk of falling infinite loop? See discussion
+     // https://github.com/tensorflow/tfjs/pull/5682#discussion_r725414008
```
[the remark this answers is not in the document]
Comment (qjia7) [reply tfjs-backend-webgpu/src/kernels/scatter_optimized_webgpu.ts]: Ack. Thanks for your clarification.

**Pull request merged**